lxcfs.c

```diff
@@ -83,12 +99,84 @@ static void users_unlock(void)
 	unlock_mutex(&user_count_mutex);
 }
 
+/*
+ * Simple functions to preserve and switch mount namespaces.
+ */
+
+/* Close all open file descriptors refering to a namespace. */
+static void close_ns(struct preserved_ns *ns)
+{
+	int i;
+	for (i = 0; i < MNT_NS_MAX; i++) {
+		if (ns->ns_fd[i] > -1) {
+			close(ns->ns_fd[i]);
+			ns->ns_fd[i] = -1;
+		}
+	}
+	ns->active_ns = -1;
+}
+
+/* Open /proc/PID/ns/mnt and save open fd to preserve the mount namespace.
+ * if @is_caller_pid is set to true it is assumed that @pid is the callers pid
+ * and that we are attached to the namespace identified by which_ns.
+ */
+static bool preserve_ns(struct preserved_ns *ns, int which_ns, int pid,
+			bool is_caller_pid)
+{
+	int ret;
+	size_t len = 5 /* /proc */ + 21 /* /int_as_str */ + 7 /* /ns/mnt */ + 1 /* \0 */;
+	char path[len];
+
+	ret = snprintf(path, len, "/proc/%d/ns/mnt", pid);
+	if (ret < 0 || (size_t)ret >= len)
+		return false;
+
+	ns->ns_fd[which_ns] = open(path, O_RDONLY | O_CLOEXEC);
+	if (ns->ns_fd[which_ns] < 0)
+		goto error;
+
+	if (is_caller_pid)
+		ns->active_ns = ns->ns_fd[which_ns];
+
+	return true;
+
+error:
+	close_ns(ns);
+	return false;
+}
+
+/* Switch caller to namespace identified by the fd retrieved via @which_ns and
+ * set the active namespace to the switched namespace. */
+static bool switch_ns(struct preserved_ns *ns, int which_ns)
+{
+	int ret = setns(ns->ns_fd[which_ns], 0);
+	if (ret < 0)
+		ns->active_ns = ns->ns_fd[which_ns] = ret;
+	else
+		ns->active_ns = ns->ns_fd[which_ns];
+
+	return ret == 0;
+}
+
+/*
+ * Functions and types used to reload dynamic library.
+ */
 static volatile sig_atomic_t need_reload;
 
 /* do_reload - reload the dynamic library.  Done under
  * lock and when we know the user_count was 0 */
-static void do_reload(void)
+static void do_reload(struct preserved_ns *ns)
 {
+	if (ns->active_ns != -1) {
```

Comment:
Also the struct preserved_ns will work as is so no need to change it, but I'm kind of wondering why you didn't just keep two global ints for the two fds.  This doesn't need to generalize right?  I'm probably missing something, hence I'm asking...